cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java

```diff
@@ -444,4 +369,209 @@ public void run() {
+    properties.put("DOMAIN", domainName + ",instance=" + instanceName);
+    helixAdmin.setConfig(scope, properties);
+  }
+
+  private static void disableInstance(
+      final HelixAdmin helixAdmin,
+      final String clusterName,
+      final String instanceId) {
+    retryWithBoundedExponentialBackOff(
+        new ExceptionalAction() {
+          @Override
+          public void apply() throws Exception {
+            LOG.error("Disabling the instance to force offlining existing resources");
+            helixAdmin.enableInstance(clusterName, instanceId, false);
+          }
+        },
+        new ExceptionalCondition() {
+          @Override
+          public boolean apply() throws Exception {
+            // When the instance is no longer enabled
+            return !helixAdmin.getInstanceConfig(clusterName, instanceId).getInstanceEnabled();
+          }
+        }, 10);
+
+    retryWithBoundedExponentialBackOff(
+        new ExceptionalAction() {
+          @Override
+          public void apply() throws Exception {
+            LOG.error(
+                "Adding \"disabled\" instance tag, to force spectator to remove this instance from "
+                    + "shard_map ");
+            helixAdmin.addInstanceTag(clusterName, instanceId, "disabled");
+          }
+        },
+        new ExceptionalCondition() {
+          @Override
+          public boolean apply() throws Exception {
+            // When the instance config tag contains the "disabled" tag.
+            return helixAdmin.getInstanceConfig(clusterName, instanceId).containsTag("disabled");
+          }
+        }, 10);
+
+    /**
+     * Here we need a mechanism to wait for all instances to go offline. However currently that logic
+     * is complicated with available java api. Hence we simply wait for 5 seconds, in the hope
+     * that the resources will be offlined in 5 seconds.
+     */
+    try {
+      Thread.currentThread().sleep(5000);
+    } catch (InterruptedException e) {
+      e.printStackTrace();
+    }
+  }
+
+  private static void enableInstance(
+      final HelixAdmin helixAdmin,
+      final String clusterName,
+      final String instanceId) {
+
+    retryWithBoundedExponentialBackOff(
+        new ExceptionalAction() {
+          @Override
+          public void apply() throws Exception {
+            LOG.error(
+                "Removing any previously added \"disabled\" instance tag,"
+                    + " to force spectator to consider this instance in shard_map "
+                    + "potentially even befoer any partitions have been active ");
+            helixAdmin.removeInstanceTag(clusterName, instanceId, "disabled");
+          }
+        },
+        new ExceptionalCondition() {
+          @Override
+          public boolean apply() throws Exception {
+            // When the instance config tag contains the "disabled" tag.
+            return helixAdmin.getInstanceConfig(clusterName, instanceId).containsTag("disabled");
```

Comment:
Should this be "when the instance config tag does not contain the "disabled" tag" since we are trying to remove the tag?

Reply:
great catch.